commands/base.go

```diff
@@ -75,6 +77,53 @@ type CommandWithContext interface {
 	RunWithContext(ctx context.Context) (Result, error)
 }
 
+// Credentials provides access to an external provider so that
+// the username, password, or agent password can be managed
+// externally, e.g. HashiCorp Vault, Ansible Vault, etc
+type Credentials struct {
+	AgentPassword string `json:"agentpassword"`
+	Password      string `json:"password"`
+	Username      string `json:"username"`
+}
+
+func (c *Credentials) Marshal() (string, error) {
+	data, err := json.Marshal(c)
+	if err != nil {
+		return "", err
+	}
+	return string(data), nil
+}
+
+func (c *Credentials) Unmarshal(jsonString string) error {
+	if err := json.Unmarshal([]byte(jsonString), &c); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Credentials) ReadFromSource(src string) error {
+	exec := false
+	if f, err := os.Lstat(src); err != nil {
+		return err
+	} else {
+		exec = f.Mode()&0111 != 0
+	}
+
+	if exec {
+		return fmt.Errorf("execution is currently unimplemented: %s is executable", src)
```

Comment:
So, that would depend upon what the policy is. However, there are already tools that are executed client-side, so it is not a blanket 'no' at this point. The thought here is that instead of needing to populate the content of a file, if execution was to be supported as an option then it would allow a script to call the user's secrets manager when the path is executable.

Reply:
@BupycHuk this has been left in place. I don't see an issue ensuring that the file is not executable even if support is never added for execution. In this scenario, it at least prevents executable files from being referenced ;)

Reply:
It should be discussed with @rnovikovP, but I think it's not a good idea to let users execute anything to get credentials. 
If we check for permissions, then error message should be different.

Reply:
```sh
pmm-admin add mysql @pmm-admin-flags.txt
failed to retrieve credentials from /usr/local/percona/pmm2/credentials.sh: execution is currently unimplemented: /usr/local/percona/pmm2/credentials.sh
```
@BupycHuk please explain what is wrong with the error message